AI/VCAI/Behaviors/RecruitHeroBehavior.cpp

```diff
@@ -33,13 +34,165 @@ Goals::TGoalVec RecruitHeroBehavior::getTasks()
 
 	if(ai->canRecruitAnyHero())
 	{
-		if(cb->getDate(Date::DAY) == 1
-			|| cb->getHeroesInfo().size() < cb->getTownsInfo().size() + 1
+		if(cb->getHeroesInfo().size() < cb->getTownsInfo().size() + 1
 			|| cb->getResourceAmount(Res::GOLD) > 10000)
 		{
 			tasks.push_back(Goals::sptr(Goals::RecruitHero()));
 		}
 	}
 
+	return tasks;
+}
+
+std::string StartupBehavior::toString() const
+{
+	return "Startup";
+}
+
+const AIPath getShortestPath(const CGTownInstance * town, const std::vector<AIPath> & paths)
+{
+	auto shortestPath = *vstd::minElementByFun(paths, [town](const AIPath & path) -> float
+	{
+		if(town->garrisonHero && path.targetHero == town->garrisonHero.get())
+			return 1;
+
+		return path.movementCost();
+	});
+
+	return shortestPath;
+}
+
+const CGHeroInstance * getNearestHero(const CGTownInstance * town)
+{
+	auto paths = ai->ah->getPathsToTile(town->visitablePos());
+
+	if(paths.empty())
+		return nullptr;
+
+	auto shortestPath = getShortestPath(town, paths);
+
+	if(shortestPath.nodes.size() > 1 
+		|| shortestPath.targetHero->visitablePos().dist2dSQ(town->visitablePos()) > 4
+		|| town->garrisonHero && shortestPath.targetHero == town->garrisonHero.get())
+		return nullptr;
+
+	return shortestPath.targetHero;
+}
+
+bool needToRecruitHero(const CGTownInstance * startupTown)
+{
+	if(!ai->canRecruitAnyHero(startupTown))
+		return false;
+
+	if(!startupTown->garrisonHero && !startupTown->visitingHero)
+		return false;
+
+	auto heroToCheck = startupTown->garrisonHero ? startupTown->garrisonHero.get() : startupTown->visitingHero.get();
+	auto paths = cb->getPathsInfo(heroToCheck);
+
+	for(auto obj : ai->visitableObjs)
+	{
+		if(obj->ID == Obj::RESOURCE && obj->subID == Res::GOLD
+			|| obj->ID == Obj::TREASURE_CHEST
+			|| obj->ID == Obj::CAMPFIRE
+			|| obj->ID == Obj::WATER_WHEEL)
```

Comment:
This should work for any `CRewardableObject` descendant

Reply:
It is interested in gold only. Maybe if I will be able to extract reward from there I can generalize it

Reply:
You can extract everything from reward configuration. 
`CRewardableObject` has `std::vector<CVisitInfo> info;` all possible rewards. Then dig into `CVisitInfo`.

Reply:
It is protected. I can not access it without extending CRewardableObject. Also I have already extended CBankInfo which itself might be unacceptable

Reply:
Just make `grantReward` & `getVisitInfo` public.